include/drivers/gpio/gpio_sim.h

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Backend API for simulated GPIO
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_GPIO_GPIO_SIM_H_
+#define ZEPHYR_INCLUDE_DRIVERS_GPIO_GPIO_SIM_H_
+
+#include <zephyr/types.h>
+#include <drivers/gpio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Simulated GPIO backend API
+ * @defgroup gpio_sim Simulated GPIO
+ * @ingroup gpio_interface
+ * @{
+ *
+ * Behaviour of virtual GPIO is application-defined. As-such, each
+ * application may
+ *
+ * - define a Device Tree overlay file to indicate the number of GPIO
+ *   controllers as well as the number of pins for each controller
+ * - register a callback with the GPIO controller using
+ *   @ref gpio_add_callback to simulate "wiring"
+ * - asynchronously call @ref gpio_sim_input_set and / or
+ *   @ref gpio_sim_input_set_masked in order to simulate GPIO events
+ *
+ * An example of an appropriate Device Tree overlay file is in
+ * tests/drivers/gpio/gpio_basic_api/boards/native_posix_64.overlay.
+ *
+ * An example of registering a callback to simulate "wiring" as well as
+ * an example of calling @ref gpio_sim_input_set is in the file
+ * tests/drivers/gpio/gpio_basic_api/src/main.c .
+ */
+
+/**
+ * @brief Modify the values of one or more simulated GPIO input @p pins
+ *
+ * @param port The simulated GPIO port
+ * @param pins The mask of pins that have changed
+ * @param values New values to assign to @p pins
+ *
+ * @return 0 on success
+ * @return -EINVAL if an invalid argument is provided
+ */
+int gpio_sim_input_set_masked(struct device *port, gpio_port_pins_t pins,
+			      gpio_port_value_t values);
+
+/**
+ * @brief Modify the value of one simulated GPIO input @p pin
+ *
+ * @param port The simulated GPIO port
+ * @param pin The pin to modify
+ * @param value New values to assign to @p pin
```

Comment:
Could you mention polarity here? I assume this is active high always

Reply:
The simulator should be operating on physical (line) levels.  Logical level translation is handled in the common API wrappers.

But yes, the API for manipulating the GPIO state external to the standard GPIO API should be clear that it uses only physical level values.  Unless we want a replicated API that works with logic level values.

Reply:
Considering this resolved.